language/bytecode-verifier/src/struct_defs.rs

```diff
@@ -73,46 +70,56 @@ impl<'a> StructDefGraphBuilder<'a> {
         }
     }
 
-    pub fn build(self) -> Graph<StructDefinitionIndex, (), Directed, u32> {
-        let mut graph = Graph::new();
-
-        let struct_def_count = self.module.struct_defs().len();
-
-        let nodes: Vec<_> = (0..struct_def_count)
-            .map(|idx| graph.add_node(StructDefinitionIndex::new(idx as TableIndex)))
-            .collect();
-
-        for idx in 0..struct_def_count {
+    pub fn build(self) -> DiGraphMap<StructDefinitionIndex, ()> {
+        let mut neighbors = BTreeMap::new();
+        for idx in 0..self.module.struct_defs().len() {
             let sd_idx = StructDefinitionIndex::new(idx as TableIndex);
-            for followed_idx in self.member_struct_defs(sd_idx) {
-                graph.add_edge(nodes[idx], nodes[followed_idx.into_index()], ());
-            }
+            self.add_struct_defs(&mut neighbors, sd_idx)
         }
 
-        graph
+        let edges = neighbors
+            .into_iter()
+            .flat_map(|(parent, children)| children.into_iter().map(move |child| (parent, child)));
+        DiGraphMap::from_edges(edges)
     }
 
-    fn member_struct_defs(
-        &'a self,
+    fn add_struct_defs(
+        &self,
+        neighbors: &mut BTreeMap<StructDefinitionIndex, BTreeSet<StructDefinitionIndex>>,
         idx: StructDefinitionIndex,
-    ) -> impl Iterator<Item = StructDefinitionIndex> + 'a {
+    ) {
         let struct_def = self.module.struct_def_at(idx);
         let struct_def = StructDefinitionView::new(self.module, struct_def);
         // The fields iterator is an option in the case of native structs. Flatten makes an empty
         // iterator for that case
-        let fields = struct_def.fields().into_iter().flatten();
-        let handle_to_def = &self.handle_to_def;
+        for field in struct_def.fields().into_iter().flatten() {
+            self.add_signature_token(neighbors, idx, field.signature_token())
+        }
+    }
 
-        fields.filter_map(move |field| {
-            let type_signature = field.type_signature();
-            let sh_idx = type_signature.token().struct_index()?;
-            match handle_to_def.get(&sh_idx) {
-                Some(sd_idx) => Some(*sd_idx),
-                None => {
-                    // This field refers to a struct in another module.
-                    None
+    fn add_signature_token(
+        &self,
+        neighbors: &mut BTreeMap<StructDefinitionIndex, BTreeSet<StructDefinitionIndex>>,
+        cur_idx: StructDefinitionIndex,
+        token: &SignatureToken,
+    ) {
+        use SignatureToken as T;
+        match token {
+            T::Bool | T::U8 | T::U64 | T::U128 | T::Address | T::TypeParameter(_) => (),
+            T::Vector(inner) | T::Reference(inner) | T::MutableReference(inner) => {
+                self.add_signature_token(neighbors, cur_idx, inner)
+            }
```

Comment:
`Reference(_) | MutableReference(_) => unreachable()`? (Assuming this runs before the signature checker?)

Reply:
I guess I could add an invariant violation, but it feels weird to since it's expressible. (definitely not a panic though)

Thoughts @dariorussi

Reply:
Ah yeah we shouldn't panic here. I'd vote for doing nothing then. It looks weird to me that we even bother looking into a malformed type.

Reply:
Conversely, to me it feels a bit weird to through an invariant violation, when the check doesn't rely on that invariant. Unsure what to do here though

Reply:
If the type is malformed it should have been caught in the previous pass and this pass wouldn't have been run at all so personally I'd still favor panicking. In general I don't think we can talk about recursive struct definitions unless the types themselves are valid.

It seems very weird to me that we even bother looking into a malformed type at all. It kinda feels like the opposite of an undefined behavior -- you have an invalid construct but you do define properties over it.

Reply:
I'm convinced. I'll add an invariant violation